### PR TITLE
Add sanakoyeu_et_al_2018 Transformerblock 

### PR DESCRIPTION
**Pull request by jbueltemeier**

### CODE REVIEW

File: pystiche_papers/sanakoyeu_et_al_2018/__init__.py
Comment (pmeier) [inline pystiche_papers/sanakoyeu_et_al_2018/__init__.py:1]:
```suggestion
from ._loss import *
```

File: pystiche_papers/sanakoyeu_et_al_2018/_loss.py
Comment (pmeier) [inline pystiche_papers/sanakoyeu_et_al_2018/_loss.py:7]:
```suggestion
__all__ = ["transformed_image_loss"]

```

**Review by pmeier — CHANGES_REQUESTED** [review]
The combination of `get_transformation_block` and `TransformerBlock` seems rather weird. Correct me if I'm wrong, but I think the `Transformer` has only two requirements:

1. It has to an instance of `pystiche.enc.Encoder`
2. It should apply `nn.AvgPool2d` or `nn.Conv2d` with respect to `impl_params`

If that is the case, I think we can do this much clearer:

```python
from pystiche import enc


class TransformerBlock(enc.SequentialEncoder):
    r"""TransformerBlock from :cite:`SKL+2018`.

    This block takes an image as input and produce a transformed image of the same size.

    Args:
        in_channels: Number of channels in the input. Defaults to ``3``.
        kernel_size: Size of the convolving kernel. Defaults to ``10``.
        stride: Stride of the convolution. Defaults to ``1``.
        padding: Padding of the input. It can be either ``"valid"`` for no padding or
            ``"same"`` for padding to preserve the size. Defaults to ``"same"``.
        impl_params: If ``True``, use the parameters used in the reference
            implementation of the original authors rather than what is described in
            the paper.

    If ``impl_params is True``, an :class:`~torch.nn.AvgPool2d` is used instead of a
    :class:`~torch.nn.Conv2d` with :func:`~torch.nn.utils.weight_norm`.
    """

    def __init__(
        self,
        in_channels: int = 3,
        kernel_size: Union[Tuple[int, int], int] = 10,
        stride: Union[Tuple[int, int], int] = 1,
        padding: str = "same",
        impl_params: bool = True,
    ):
        kwargs = {
            "kernel_size": kernel_size,
            "stride": stride,
            "padding": get_padding(padding, kernel_size),
        }
        module = (
            nn.AvgPool2d(**kwargs)
            if impl_params
            else nn.utils.weight_norm(nn.Conv2d(in_channels, in_channels, **kwargs))
        )
        super().__init__((module,))
```

Thoughts?

File: pystiche_papers/sanakoyeu_et_al_2018/_loss.py
Comment (pmeier) [inline pystiche_papers/sanakoyeu_et_al_2018/_loss.py:31]:
```suggestion
        score_weight = 1e2 if impl_params else 1e0
```

File: pystiche_papers/sanakoyeu_et_al_2018/_modules.py
Comment (pmeier) [inline pystiche_papers/sanakoyeu_et_al_2018/_modules.py:224]: Why do you need an `ignore[arg-type]` here? What is the error message? Is it maybe enough to declare `module: nn.Module = ...`?

File: pystiche_papers/sanakoyeu_et_al_2018/_modules.py
Comment (pmeier) [inline pystiche_papers/sanakoyeu_et_al_2018/_modules.py:231]: Why do we need padding here? Isn't that handled by the module itself?

File: pystiche_papers/sanakoyeu_et_al_2018/_modules.py
Comment (pmeier) [inline pystiche_papers/sanakoyeu_et_al_2018/_modules.py:235]: If we need that extra padding for some reason that could be made more robust by 
```suggestion
        return super().forward(input)
```

**Review by pmeier — CHANGES_REQUESTED** [review]
Test failures are related. Please address them in addition to my comments.

File: pystiche_papers/sanakoyeu_et_al_2018/_modules.py
Comment (jbueltemeier) [reply pystiche_papers/sanakoyeu_et_al_2018/_modules.py:231]: The size of the image will be changed otherwise, but I think this should be integrated in `same_size_padding` and not here. It is probably due to the `kernel_size=10`.

File: pystiche_papers/sanakoyeu_et_al_2018/_modules.py
Comment (jbueltemeier) [reply pystiche_papers/sanakoyeu_et_al_2018/_modules.py:224]: > pystiche_papers\sanakoyeu_et_al_2018\_modules.py: note: In member "__init__" of class "TransformerBlock":
pystiche_papers\sanakoyeu_et_al_2018\_modules.py:224: error: Argument 1 to "AvgPool2d" has incompatible type
"**Dict[str, Union[Tuple[int, int], int]]"; expected "bool"  [arg-type]
                nn.AvgPool2d(**kwargs)

At nn. Conv2d are similar error messages.

Comment (pmeier) [reply pystiche_papers/sanakoyeu_et_al_2018/_modules.py:224]: Hm, OK I didn't think of that. In that case, I think it is OK as is.

File: pystiche_papers/sanakoyeu_et_al_2018/_modules.py
Comment (pmeier) [reply pystiche_papers/sanakoyeu_et_al_2018/_modules.py:231]: Yeah, that means `same_size_padding` does not work as expected. Could you fix this in another PR and rebase this after the other is merged?

Comment (pmeier) [reply pystiche_papers/sanakoyeu_et_al_2018/_modules.py:231]: Fixed in #175

**Comment by jbueltemeier** [comment]
We need the `SameSizeConv2D` fixed in [#175](https://github.com/pmeier/pystiche_papers/pull/175) also for an `nn.AvgPool2D`.

**Comment by pmeier** [comment]
> We need the `SameSizeConv2D` fixed in [#175](https://github.com/pmeier/pystiche_papers/pull/175) also for an `nn.AvgPool2D`.

I'm on it.

**Comment by pmeier** [comment]
#179 adds the functionality that you need.

**Review by pmeier — APPROVED** [review]
@jbueltemeier This is good to merge after the conflicts are resolved.

**Pull request merged**
